File: src/main/java/com/voxel/World.java
package com.voxel;

/**
 * Simple block grid world. Stores blocks as bytes:
 * 0 = AIR
 * 1 = DIRT
 * 2 = GRASS
 */
public class World {
    public static final byte AIR = 0;
    public static final byte DIRT = 1;
    public static final byte GRASS = 2;

    private final int width;
    private final int height;
    private final int depth;
    private final byte[] blocks;

    public World(int width, int height, int depth) {
        this.width = width;
        this.height = height;
        this.depth = depth;
        this.blocks = new byte[width * height * depth];
        generateFlatWorld();
    }

    private int idx(int x, int y, int z) {
        return x + z * width + y * (width * depth);
    }

    private void setBlock(int x, int y, int z, byte id) {
        if (inBounds(x, y, z)) blocks[idx(x, y, z)] = id;
    }

    public byte getBlock(int x, int y, int z) {
        if (!inBounds(x, y, z)) return AIR;
        return blocks[idx(x, y, z)];
    }

    private boolean inBounds(int x, int y, int z) {
        return x >= 0 && x < width && y >= 0 && y < height && z >= 0 && z < depth;
    }

    private void generateFlatWorld() {
        int ground = Math.max(1, height / 4); // small ground height
        for (int x = 0; x < width; x++) {
            for (int z = 0; z < depth; z++) {
                for (int y = 0; y < height; y++) {
                    if (y < ground) {
                        setBlock(x, y, z, DIRT);
                    } else {
                        setBlock(x, y, z, AIR);
                    }
                }
                // top surface becomes grass
                setBlock(x, ground - 1, z, GRASS);
            }
        }
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getDepth() {
        return depth;
    }
}

File: js/world.js
// World.js - Manages the voxel world
export class World {
    static AIR = 0;
    static DIRT = 1;
    static GRASS = 2;

    constructor(width, height, depth) {
        this.width = width;
        this.height = height;
        this.depth = depth;
        this.blocks = new Uint8Array(width * height * depth);
        this.generateFlatWorld();
    }

    idx(x, y, z) {
        return x + z * this.width + y * (this.width * this.depth);
    }

    setBlock(x, y, z, id) {
        if (this.inBounds(x, y, z)) {
            this.blocks[this.idx(x, y, z)] = id;
        }
    }

    getBlock(x, y, z) {
        if (!this.inBounds(x, y, z)) return World.AIR;
        return this.blocks[this.idx(x, y, z)];
    }

    inBounds(x, y, z) {
        return x >= 0 && x < this.width && 
               y >= 0 && y < this.height && 
               z >= 0 && z < this.depth;
    }

    generateFlatWorld() {
        const ground = Math.max(1, Math.floor(this.height / 4));
        for (let x = 0; x < this.width; x++) {
            for (let z = 0; z < this.depth; z++) {
                for (let y = 0; y < this.height; y++) {
                    if (y < ground) {
                        this.setBlock(x, y, z, World.DIRT);
                    } else {
                        this.setBlock(x, y, z, World.AIR);
                    }
                }
                // Top surface becomes grass
                this.setBlock(x, ground - 1, z, World.GRASS);
            }
        }
    }
}
